src/TheSpace/HarbergerMarket.sol

```diff
@@ -69,120 +84,177 @@ abstract contract HarbergerMarket is Multicall {
 
-        // collect tax
-        bool success = this.collectTax(tokenId_);
+            if (success) {
+                // successfully clear tax
+                currency.transferFrom(msg.sender, property.ownerOf(tokenId), askPrice);
+                property.safeTransferByMarket(property.ownerOf(tokenId), msg.sender, tokenId);
 
-        if (success) {
-            // successfully clear tax
-            currency.transferFrom(msg.sender, property.ownerOf(tokenId_), price_);
-            property.safeTransferByMarket(property.ownerOf(tokenId_), msg.sender, tokenId_);
-        } else {
-            // if failed, mint to current bidder
-            property.mint(msg.sender, tokenId_);
+                return;
+            }
         }
+
+        // if token does not exists yet, or token is defaulted
+        // mint token to current sender for free
+        property.mint(msg.sender, tokenId);
+        // update tax record
+        taxRecord[tokenId].lastTaxCollection = block.number;
     }
 
     /**
      * @dev Collect outstanding property tax for a given token, put token on tax sale if obligation not met.
      *
      * Emits a {Tax} event and a {Price} event (when properties are put on tax sale).
      */
-    function collectTax(uint256 _tokenId) external returns (bool) {
-        uint256 price = this.getPrice(_tokenId);
-        if (price > 0) {
-            // TODO: time window for tax rate
-
-            // calculate tax
-            uint256 tax = (price * taxRate * (block.timestamp - taxRecord[_tokenId].lastTaxCollection)) / 100;
+    function collectTax(uint256 tokenId) external returns (bool) {
+        if (!property.exists(tokenId)) revert TokenNotExists();
 
+        uint256 tax = this.getTax(tokenId);
+        if (tax > 0) {
             // calculate collectable amount
-            address taxpayer = property.ownerOf(_tokenId);
+            address taxpayer = property.ownerOf(tokenId);
             uint256 allowance = currency.allowance(taxpayer, address(this));
             uint256 balance = currency.balanceOf(taxpayer);
-
             uint256 collectable = _min(allowance, balance);
 
-            // collect tax or default
-            if (tax < collectable) {
+            // calculate amount to be collected, the smaller one of tax and collectable
+            // then update accumulatedUBI
+            uint256 collecting = _min(collectable, tax);
+
+            if (collecting > 0) {
+                currency.transferFrom(property.ownerOf(tokenId), address(this), collecting);
+                emit Tax(tokenId, collecting);
+                // update accumulated ubi
+                accumulatedUBI += collecting;
+            }
+
+            // default if tax is not fully collected
+            if (tax > collectable) {
                 // default
-                currency.transferFrom(property.ownerOf(_tokenId), address(this), collectable);
-                _default(_tokenId);
+                _default(tokenId);
```

Comment:
What happens on taxReocrd of this pixel when `_default` a token? Will it be set to `block.number`?

Reply:
`lastTaxCollection` is not updated here, it will be updated when the token is minted again [here](https://github.com/thematters/contracts/pull/27/files?diff=split&w=0#diff-c8d4530178144324afe66c9ad98e27c0423239a6b4da88ba6b2d3e06c9ebdc13R235), so the tax will not accumulate.